pkg/insightstore/report.go

```diff
@@ -0,0 +1,250 @@
+// Copyright 2020 The PipeCD Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package insightstore
+
+import (
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/pipe-cd/pipe/pkg/model"
+)
+
+// deploy frequency
+
+// DeployFrequencyReport satisfy the interface `Report`.
+type DeployFrequencyReport struct {
+	AccumulatedTo int64                    `json:"accumulated_to"`
+	Datapoints    DeployFrequencyDataPoint `json:"datapoints"`
+	FilePath      string
+}
+
+type DeployFrequencyDataPoint struct {
+	Daily   map[string]DeployFrequency `json:"daily"`
+	Weekly  map[string]DeployFrequency `json:"weekly"`
+	Monthly map[string]DeployFrequency `json:"monthly"`
+	Yearly  map[string]DeployFrequency `json:"yearly"`
+}
+
+type DeployFrequency struct {
+	DeployCount float32 `json:"deploy_count"`
+}
+
+func (d *DeployFrequencyReport) GetFilePath() string {
+	return d.FilePath
+}
+
+func (d *DeployFrequencyReport) PutFilePath(path string) {
+	d.FilePath = path
+}
+
+func (d *DeployFrequencyReport) Value(step model.InsightStep, key string) (float32, error) {
```

Comment:
For example, in the case of wanting the daily data points of a month, we use this function?

Reply:
Partially yes.

The only time we want to get data for multiple days is when we want to get InsightDataPoints, which is data for a graph

So, when
> in the case of wanting the daily data points of a month
 I think that we use convertToInsightDataPoints (convertToInsightDataPoints uses Value internally, though.)